util/limiter/limiter.go

```diff
@@ -0,0 +1,124 @@
+package limiter
+
+import (
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru/v2/expirable"
+)
+
+// Limiter is a keyed token bucket rate limiter.
+//
+// Each key gets its own separate token bucket to pull from, enabling
+// enforcement on things like "requests per IP address". To avoid
+// unbounded memory growth, Limiter actually only tracks limits
+// precisely for the N most recently seen keys, and assumes that
+// untracked keys are well-behaved. This trades off absolute precision
+// for bounded memory use, while still enforcing well for outlier
+// keys.
+//
+// As such, Limiter should only be used in situations where "rough"
+// enforcement of outliers only is sufficient, such as throttling
+// egregious outlier keys (e.g. something sending 100 queries per
+// second, where everyone else is sending at most 5).
+//
+// Each key's token bucket behaves like a regular token bucket, with
+// the added feature that a bucket's token count can optionally go
+// negative. This implements a form of "cooldown" for keys that exceed
+// the rate limit: once a key starts getting denied, it must stop
+// requesting tokens long enough for the bucket to return to a
+// positive balance. If the key keeps hammering the limiter in excess
+// of the rate limit, the token count will remain negative, and the
+// key will not be allowed to proceed at all. This is in contrast to
+// the classic token bucket, where a key trying to use more than the
+// rate limit will get capped at the limit, but can still occasionally
+// consume a token as one becomes available.
+type Limiter[K comparable] struct {
+	// Size is the number of keys to track. Only the Size most
+	// recently seen keys have their limits enforced precisely, older
+	// keys are assumed to not be querying frequently enough to bother
+	// tracking.
+	Size int
+	// TTL is an optional timeout for tracked keys. Keys not used for
+	// TTL are removed from the limiter's state, even if they fall
+	// within the Size most recently used. If TTL is zero, up to Size
+	// keys are tracked regardless of how long they've sat unused.
+	TTL time.Duration
+
+	// Burst is the number of tokens available for a key to consume
+	// before time-based rate limiting kicks in. An unused limiter
+	// regains available tokens over time, up to Max tokens. A newly
+	// tracked key initially receives Max tokens.
+	Max int64
+	// Overdraft is the amount of additional tokens a key can consume
```

Comment:
It might be worth distinguishing "consume" here, which implies we are giving them a lease to query. Perhaps: "Overdraft is a number of additional tokens a key may be charged for when it exceeds its rate limit. Each additional request issued for the key charges one unit of overdraft, up to this limit. Overdraft tokens are refilled at the normal rate, and must be fully repaid before any tokens will be made available for requests." or words to that effect.

Reply:
Done